File: src/linked-list/python3/21_MergeTwoSortedLists.py
#
# @lc app=leetcode id=21 lang=python3
#
# [21] Merge Two Sorted Lists
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        result_node=ListNode()
        cur_node=result_node
        while list1 or list2:
            if not list1:
                cur_node.next=list2
                break
            elif not list2:
                cur_node.next=list1
                break
            else:
                if list1.val <= list2.val:
                    cur_node.next=ListNode(val=list1.val)
                    list1=list1.next
                else:
                    cur_node.next=ListNode(val=list2.val)
                    list2=list2.next
                cur_node=cur_node.next
        
        return result_node.next
        
# @lc code=end

File: work/typescript/linked-list/21_MergeTwoSortedLists.ts
/*
 * @lc app=leetcode id=21 lang=typescript
 *
 * [21] Merge Two Sorted Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
    const result:ListNode = new ListNode();
    let resultTmp:ListNode = result;
    while(list1 && list2){
        if(list1.val < list2.val){
            resultTmp.next = new ListNode(list1.val);
            list1=list1.next;
        }
        else{
            resultTmp.next = new ListNode(list2.val);
            list2=list2.next;
        };
        resultTmp=resultTmp.next;
    };
    if(list1){
        resultTmp.next=list1;
    };
    if(list2){
        resultTmp.next=list2;
    };
    return result.next;
}
// @lc code=end

